src/failure_message.rs

```diff
@@ -76,40 +101,114 @@ pub struct MessagePayload {
+}
+
+impl TemplateFailure {
+    fn new(
+        test_suite: String, 
+        test_name: String, 
+        duration: String, 
+        raw_num_backticks: usize, 
+        build_url: Option<String>,
+        stack_trace: Vec<String>
+    ) -> Self {
+        let num_backticks = max(raw_num_backticks + 1, 3);
+        let backticks = String::from("`".repeat(num_backticks));
+        Self { test_suite, test_name, duration, backticks, build_url, stack_trace }
+    }
+}
+
+fn longest_repeated_substring(s: String, target: char) -> usize {
+    let mut max_length = 0;
+    let mut current_length = 0;
+
+    for c in s.chars() {
+        if c == target {
+            current_length += 1;
+            max_length = max_length.max(current_length);
+        } else {
+            current_length = 0; // Reset when the character doesn't match
+        }
+    }
+
+    max_length
+}
+
 #[pyfunction]
 pub fn build_message(payload: MessagePayload) -> String {
-    use std::fmt::Write;
-    let mut message = String::from("### :x: Failed Test Results:\n");
+    let mut tera = Tera::default();
+    tera.add_raw_template("test_results_template", TEMPLATE).unwrap();
 
     let failed: i32 = payload.failed;
     let passed: i32 = payload.passed;
     let skipped: i32 = payload.skipped;
 
     let completed = failed + passed + skipped;
-    writeln!(&mut message, "Completed {completed} tests with **`{failed} failed`**, {passed} passed and {skipped} skipped.").unwrap();
-    message.push_str("<details><summary>View the full list of failed tests</summary>\n\n");
-    message.push_str("| **Test Description** | **Failure message** |\n");
-    message.push_str("| :-- | :-- |\n");
-
-    for (idx, fail) in payload.failures.into_iter().enumerate() {
-        if idx != 0 {
-            message.push('\n');
-        }
-        message.push_str("| <pre>");
-        write!(
-            &mut message,
-            "Testsuite:<br>{}<br><br>Test name:<br>{}<br>",
-            fail.testsuite, fail.name
-        )
-        .unwrap();
-        message.push_str("</pre> | <pre>");
-
-        match fail.failure_message {
-            None => message.push_str("No failure message available"),
-            Some(x) => message.push_str(&escape_failure_message(&shorten_file_paths(&x))),
-        }
-        message.push_str("</pre> |");
-    }
 
+    let mut sorted_failures: Vec<Failure> = payload.failures.to_vec();
+    sorted_failures.sort_by(|a, b| a.duration.partial_cmp(&b.duration).unwrap());
+
+    let mut template_failures: Vec<TemplateFailure> = Vec::new();
+    sorted_failures.truncate(3);
+    for failure in sorted_failures.iter_mut() {
+        let failure_message = match failure.failure_message.as_ref() {
+            Some(x) => String::from(x),
+            _ => String::from("No failure message available"),
+        };
+        let stack_trace_lines: Vec<String> = failure_message
+            .split('\n')
+            .map(|s| escape_message(s).to_string())
+            .collect();
```

Comment:
```suggestion
        let failure_message = failure.failure_message.map(|s| s.as_str()).unwrap_or("No failure message available");
        let stack_trace_lines: Vec<_> = failure_message
            .split('\n')
            .map(escape_message)
            .collect();
```

`escape_message` already returns a `String`, so the `to_string` is redundant here.
You can also avoid cloning the `failure_message` by using `as_str`, and the manual match by using `unwrap_or`.